Publish `send` module tutorial

jbesraa opened the pull request:
The example implements the sender part of a pj transaction and sets up
  a receiver through payjoin-cli to complete the transaction.

jbesraa commented:
waiting for  https://github.com/payjoin/rust-payjoin/issues/189

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,403 @@
+
+    let sender_client = bitcoincore_rpc::Client::new(
+        "http://localhost:18443/wallet/sender",
+        Auth::UserPass(
+            "foo".to_string(),
+            "qDDZdeQ5vw9XXFeVnXT4PZ--tGN2xNjjR4nrtyszZx0=".to_string(),
+        ),
+    )
+    .unwrap();
+    let receiver_client = bitcoincore_rpc::Client::new(
+        "http://localhost:18443/wallet/receiver",
+        Auth::UserPass(
+            "foo".to_string(),
+            "qDDZdeQ5vw9XXFeVnXT4PZ--tGN2xNjjR4nrtyszZx0=".to_string(),
+        ),
```
Can auth be saved in a variable so as not to repeat it?

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,403 @@
+  
+const LOCAL_CERT_FILE: &str = "localhost.der";
+
+fn https_agent() -> ureq::Agent {
+    use ureq::AgentBuilder;
+
+    let mut local_cert_path = PathBuf::from("/tmp");
+    local_cert_path.push(LOCAL_CERT_FILE);
+    let cert_der = std::fs::read(local_cert_path).unwrap();
+    let mut root_cert_store = RootCertStore::empty();
+    root_cert_store.add(&Certificate(cert_der)).unwrap();
+    let client_config = ClientConfig::builder()
+        .with_safe_defaults()
+        .with_root_certificates(root_cert_store)
+        .with_no_client_auth();
+
+    AgentBuilder::new()
+        .tls_config(Arc::new(client_config))
+        .build()
+}
```
Payjoin Dev Kit uses ureq in order to support msrv, but it's a less familiar crate than `reqwest`. One should be able to use a `reqwest::blocking::Client` with [`danger_accept_invalid_certs`](https://docs.rs/reqwest/latest/reqwest/struct.ClientBuilder.html#method.danger_accept_invalid_certs) to demonstrate this example without reading from a local cert in a file every run

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,403 @@
+            .is_ok());
+    }
+
+    let sender_client = bitcoincore_rpc::Client::new(
+        "http://localhost:18443/wallet/sender",
+        Auth::UserPass(
+            "foo".to_string(),
+            "qDDZdeQ5vw9XXFeVnXT4PZ--tGN2xNjjR4nrtyszZx0=".to_string(),
+        ),
+    )
+    .unwrap();
+    let receiver_client = bitcoincore_rpc::Client::new(
+        "http://localhost:18443/wallet/receiver",
+        Auth::UserPass(
+            "foo".to_string(),
+            "qDDZdeQ5vw9XXFeVnXT4PZ--tGN2xNjjR4nrtyszZx0=".to_string(),
```
Auth and the url should be saved in variables so they don't need to be repeated

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
````diff
@@ -0,0 +1,403 @@
+ following command*
+
+
+ ```
+ docker run --name bitcoin-server -d --rm -it \
+ -p 18443:18443 \
+ -p 18444:18444 \
+ ruimarinho/bitcoin-core \
+ -printtoconsole \
+ -regtest=1 \
+ -rpcallowip=0.0.0.0/0 \
+ -rpcbind=0.0.0.0 \
+ -rpcauth='foo:7d9ba5ae63c3d4dc30583ff4fe65a67e$9e3634e81c11659e3de036d0bf88f89cd169c1039e6e09607562d54765c649cc'
+ ```
````
Docker might be able to reckon with to write a tutorial, but it's more work for the person following the tutorial and leaves side-effects on their system. I'd much rather we use the [`bitcoind` crate](https://docs.rs/bitcoind/latest/bitcoind/) to handle that setup and tear down for them.

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,403 @@
+in a payjoin transaction and transact with a payjoin receiver
+utilising `payjoin-cli`.
+
+While `rust-payjoin` supports both BiP0078 and BiP0077, this tutorial
```
Any time a spec is referenced it's got to be hyperlinked

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,403 @@
+crate to run a regtest node and fund two wallets, one for the sender
+and one for the receiver.
+
+For any feedback, please free to reach out through the payjoin
+official discord or github.
```
Any call to action should also be completed with hyperlinks. Don't make the reader do more work than strictly necessary. This should be a welcoming tutorial after all

DanGould commented:
While this example may work I'm asking that it be cleaned up for ergonomics of the reader.

In addition to my prior review and copy edits, we should be using default bitcoind cookie auth instead of legacy user:password auth too.

DanGould commented:
I see now why bitcoind may not have seemed easy for such an example: it would require payjoin-cli also be run by the example so that it can get the connection details of bitcoind in the same process. While this may be a bit more complex, I think it's actually Ideal since it doesn't leave side effects on the reader's box or require that they have docker installed.

Fortunately, this code has already been [written in the integration test suite](https://github.com/payjoin/rust-payjoin/blob/bc8c83af749f89867545e637abc4479fae1141b8/payjoin/tests/integration.rs#L641). Rather than have the reader start with a blank cargo project, it might make sense to host one in a git repo they can clone and then iterate on. That way they don't have to concern themself at all with the setup or wallet funding and can focus entirely on the tutorial of writing a Payjoin Dev Kit sender.

Let me explain in detail:

1. User git clones a repo containing an outline'd rust file including the bitcoind and payjoin-cli fixtures
2. User follows tutorial
3.  User runs the working example and can compare it to an answer file, also contained in the example repo, including a completed sender as outlined in the tutorial. This could even be a specific commit of a running example in the main rust-payjoin repository

jbesraa commented:
While I agree that it would be nice to have a FULLY contained example, I am not strongly convinced that the current solution is not good enough.
0) good chunk of the people who will try the tutorial will probably know how to run bitcoin core, and maybe they will have one running already, if we inject the wallet+bitcoind, those users will be required to actually change the code they clone. 
1) the current solution is not presenting too much fraction, running a single docker command shouldn't be complicated for someone who is trying to understand payjoin?
2) while this example is demonstrating how to implement the sender, we do also demonstrate how to run the cli receiver, if we move it to code, the user wont get the benefit of understanding the cli as well
3) bitcoind client is not implementing some traits that I think we would need(sync/send/ ?)

I agree about the "footprint" we leave on the user machine, ill add another step for the user to clean that. 

To follow the 80/20 rule, I think the current wallet solution is by far meet this rule, and because we are not pushing here production code or something very crucial, I think its good. If we release the tutorial and people will be like "woo what is this wallet thing" we will know we(I) was wrong about it. 

Overall, at this point, I don't believe that the benefit of having a fully contained wallet is convincing.

jbesraa commented:
Changed ureq to reqwest and other small changes. I think its ready for merge after a review of the language(typos and so) and fixing all the hyperlinks.

DanGould commented:
80/20 of getting something across the line that significantly improves on the existing docs is a sound line of thought.
My original 80/20 conception was to just take the existing docs and post them **AS-IS** in an `.md` on the website. We agreed on an example, and I made this suggestion in discord so we didn't lose an already functional document.

See discord: "put [the removed docs] as-is into a .md until we have a working example and don't just lose docs. Happy to host on payjoindevkit.org"

This PR is making the same scope-expanding issue as PjUriBuilder. If it were just a spruced up MD of the text on the top of the docfile I wouldn't give it the same scrutiny. 

> While I agree that it would be nice to have a FULLY contained example, I am not strongly convinced that the current solution is not good enough.

"good enough" is what we had as is. This is neither an attempt to produce the contained examples we discussed when we last met nor an as-is copy. It's a new, third thing, a tutorial on payjoindevkit.org. 

What you have here should still be a solid base toward a working, contained send side [example target](https://doc.rust-lang.org/cargo/reference/cargo-targets.html?highlight=example#examples).

Let's focus on the direction we already agreed upon unless something fundamental changed from that conversation until now. I also re-opened [the example issue], because we both planned on including examples of running code last we met together. 

--

I'm going to address your direct concerns, but I'd really just like to see those docs that were deleted incoroporated into payjoindevkit.org as is so we can get working, isolated examples, as planned

>     1. good chunk of the people who will try the tutorial will probably know how to run bitcoin core, and maybe they will have one running already, if we inject the wallet+bitcoind, those users will be required to actually change the code they clone.

Since the current state of this PR is a tutorial and not a static example, changing code is the point! As stated above, I don't think this approach actually the most pressing thing to write at the moment.

>     2. the current solution is not presenting too much fraction, running a single docker command shouldn't be complicated for someone who is trying to understand payjoin?

Having tutored quite a few people through payjoin-cli usage, I think it is too much friction, especially for a green-field publication where we can get it right the first time. I'd rather not assume docker install or knowledge.

>     3. while this example is demonstrating how to implement the sender, we do also demonstrate how to run the cli receiver, if we move it to code, the user wont get the benefit of understanding the cli as well

Focused, isolated tutorials are easier to follow. *If* we want a cli tutorial beyond the payjoin-cli readme, it should be separate, focused and isolated.

>     4. bitcoind client is not implementing some traits that I think we would need(sync/send/ ?)

I don't think an example necessitates `async` or a tokio runtime, which is why the compiler wants Sync/Send. Avoid that by using blocking (with reqwest::blocking and Rouille) to avoid sync/send requirement. Alternatively, just use tokio::task::spawn_blocking to manage blocking code in a tokio runtime.

The work done here is not in vain! I think you've just trod a bit beyond the direction we set out on and we should stick to making progress on the documented plan unless fundamental needs have changed, in which case we should document those and change course. I'll paste the plan from the meeting notes that I left in discord for easy access

> ## Payjoin Example / docs
> 
> - remove non-compiling docs but not all mod docs
> - address just the one concern in the PR
> - 🆕 put them as-is into a .md until we have a working example and don't just lose docs. Happy to host on payjoindevkit.org
> - Isolate sender / receiver examples (start with send imo) that stand up bitcoind and payjoin-cli as counterparty to show each distinct example in isolation

jbesraa commented:
aye, iam fairly confused here..this is added under tutorials to go here https://payjoindevkit.org/tutorials/getting-started/ 
and now we are suggesting different stuff that were dismissed before..

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
````diff
@@ -0,0 +1,372 @@
+
+And lets call the sender functions in the main function:
+
+```
````
```rust

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
````diff
@@ -0,0 +1,372 @@
+ 1. Install `payjoin-cli`
+
+ ```shell
+ cargo install payjoin-cli --version 0.0.2-alpha --features local-danger-https
````
/local-danger-https/danger-local-https/

This is failing due to rustls version mismatch in the danger-local-https deps. We'll have to release a fix this before the tutorial ships. I'll proceed using a locally built payjoin-cli.

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
````diff
@@ -0,0 +1,372 @@
+ successfully and generated the `receiver` wallet*.
+
+ ```rust
+  payjoin-cli -r http://localhost:18443/wallet/reciever --rpcuser foo --rpcpass qDDZdeQ5vw9XXFeVnXT4PZ--tGN2xNjjR4nrtyszZx0= receive --endpoint https://localhost:3000 1230000000
````
A few things could simplify this manual setup

    Writing a script that automatically runs a payjoin-cli receiver in the docker container
    using a simple password like "insecurepassword"
    using the default cookie auth and having the container output the cookie location on startup

Looks like the script also created a wallet called receiver while this command wants a wallet called reciever (the former is common english)

remove the indentation

DanGould reviewed (commented):
Nice! following this tutorial I was able to make a successful payjoin, but It left mea wanting to know more about what code I wrote and why. I could just paste 2 chunks of code and press play without knowing why PDK does what it does

I think it's valuable to keep in mind [what this was intended to improve upon](https://docs.rs/payjoin/latest/payjoin/send/index.html), which included step-by-step commentary for each implementation detail.

It would be nice if the tutorial focused on breaking down each step of the sender's role piece by piece. Zavior appears to do that in his [flutter tutorial](https://bitcoindevkit.org/blog/exploring-bdk-flutter/) where he breaks down each new section in contrast to an example, which is where I know this was headed and probably why it exists in current form.

Most of the tutorial seems to be centered on environment setup, which even I ran into some snags with. Ideally we can automate away. I'm glad you changed to reqwest, which makes dealing with cert validation way easier for a tutorial like this. I don't want to get too bogged down in automation, but I'd like for the content of the Payjoin Dev Kit sender tutorial to be the majority of the article. Would publishing our own a docker container with payjoin-cli running on, regtest bitcoind, with 2 wallets set up be a happy medium that would let the instructions here largely same the same and focus to be on the Payjoin Dev Kit part?

jbesraa commented:
@DanGould thanks for the review.

following your two main points about breaking down the sender and the node/wallets setup, I made some changes and would appreciate your review. 

For setting up the wallet/node, I published a crate `pjregtest` that is starting bitcoin node in regtest and it creates funded wallets, so the user now just `cargo install pjregtest && pjregtest` and they will have the node up with `sender` and `receiver` wallets.
For the sender part, I broke things down. on thing that iam not sure about is either I should keep the function signature in each step or remove it. from one perspective it gives the user context from the other side, you cant copy the whole code block as it will be wrong.

I added an intro paragraph charged with some ai =) 

didnt yet fix all the typos and so, please lets finish reviewing the code structure/steps and ill focus on the wording after that.

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,311 @@
+# Privacy First: Send Payjoin Transactions with Payjoin Dev Kit
+
+Want to enhance your Bitcoin transactions with enhanced privacy and reduced
+fees? Look no further than Payjoin and the powerful rust-payjoin crate! Payjoin
+allows you to combine your transaction with others, minimizing individual
+footprints and potentially saving on network fees.
+
+This tutorial will guide you through the exciting world of Payjoin using Rust.
+We'll delve into the rust-payjoin library, providing step-by-step instructions
+on crafting and sending your very own Payjoin transactions. Moreover, We'll
+guide you through setting up payjoin-cli, and use it a the receiver side in the
+payjoin transaction.
```
Pleasantly surprised by the prose generated here. I think we still want to focus on how it's the `payjoin` crate (since that's how it exists on crates.io) and `send` module.

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
````diff
@@ -0,0 +1,311 @@
+payjoin transaction.
+
+# Bitcoin Node & Wallets
+  In order to perform a payjoin transaction, the following code will create two
+  bitcoin wallets, "sender" and "receiver" connected to bitcoin node running in
+  regtest mode.
+
+  * You need to have docker installed!
+
+  ```shell
+  cargo install pjregtest
+  ```
+
+  ```shell
+  pjregtest
+  ```
````
What is up with this indentation that keeps appearing? It continues later in the file

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,311 @@
+  bitcoin wallets, "sender" and "receiver" connected to bitcoin node running in
+  regtest mode.
+
+  * You need to have docker installed!
```
installed and daemon running*

Use subheadings `##...` instead of bullets for subheadings or `**...**` bold for emphasis sparingly. I like to keep exclamation points to exclamations but that's just style.

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
````diff
@@ -0,0 +1,311 @@
+
+  * You need to have docker installed!
+
+  ```shell
+  cargo install pjregtest
+  ```
````
Congrats on publishing your first crate

where's the source code? I don't think we want to get our users in the habit of running arbitrary code. We

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
````diff
@@ -0,0 +1,311 @@
+    `try_payjoin`. The function will take a `payjoin::Uri` as an input and
+    attempt to perform a payjoin transaction.
+    
+```rust
+struct Sender;
+
+impl Sender {
+    pub async fn try_payjoin(pj_uri: payjoin::Uri<'static, NetworkChecked>) -> Txid {
+    }
+}
+```
````
if the sender doesn't hold state, you're better off either making a `mod sender` instead of a struct or just leaving the functions in the main body of the file.

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,311 @@
+guide you through setting up payjoin-cli, and use it a the receiver side in the
+payjoin transaction.
+
+# Bitcoin Node & Wallets
```
Only the title should be h1 (`#`), subheadings should use `##`, `###` etc.

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,311 @@
+`try_payjoin` function.
+
+* Extract the parameters from the payjoin URI
+
+  Here we extract the amount the reciever is requesting and the address they
+  wish to recieve to. The URI also includes the endpoint we will make the
+  request to, and that is extracted in the `RequestContext::from_psbt_and_uri`
+  function and returned as part of the request data.
```
explain why, why why.

Some thing like "Payjoin requires messages to be sent between sender and receiver. The receiver hosts an HTTP endpoint to receive these messages. To make a request, we need a payjoin Uri which we will parse in the following function:"

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,311 @@
+* Construct an "Original PSBT" paying `.amount` to `.address` as specified in
+  the URI.
+
+  First we add the reciever address to the outputs list along side the request
+  amount. Then we create a funded psbt with options set to specify `fee_rate`
+  and other paramters.
```
explain why we're making an original psbt and how it's just a fallback in case we disappear

jbesraa replied:
sorry, can u please explain further? which side would disappear?

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
````diff
@@ -0,0 +1,311 @@
+  and other paramters.
+
+```rust
+...
+pub async fn try_payjoin(pj_uri: payjoin::Uri<'static, NetworkChecked>) -> Txid {
+    ...
````
comment out `...` so it's not so easy to copy an error

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
````diff
@@ -0,0 +1,311 @@
+}
+```
+
+* Construct the HTTPS request
+
+  Prepare the request we want to send and also get the context back so we can
+  handle the response when its recieved.
````
This is a good opportunity to explain why authenticated communication is necessary and plaintext is insufficient.

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
````diff
@@ -0,0 +1,311 @@
+
+
+```rust
+#[tokio::main]
````
You can use `reqwest::blocking::Client` to avoid the complexity of an async runtime with reqwest's `blocking` feature

DanGould reviewed (commented):
The focus of this tutorial is far more concise and feels more like a tutorial for the actual sending of a payjoin.

The source code for both the pjregtest should be open source and available so that we don't encourage users to run trusted binaries, even if this is only a test. I still prefer a docker container on dockerhub since it's contained and can easily be deleted, but I'm not sure what pjregtest is doing (maybe something contained?)

As far as your concern for a complete copy paste, I suggest hosting a completed example in a gist that you link at the end as a sort of solution file that can be used to debug. Overall this makes it easy to see how easy it is to send a payjoin with PDK. I like it

jbesraa commented:
@DanGould Thank you so much for the review.

I believe I went through all of your comments and things look much cleaner now. I really want to test this with published `cli` so once we have that ill go through this again to fix the text for the cli and test it with it.

jbesraa commented:
Tested with cli 0.0.4-alpha and all works as expected.
I think that covers all of it for this pr.

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,359 @@
+// ...
+// pub fn try_payjoin(pj_uri: payjoin::Uri<'static, NetworkChecked>) -> Txid {
+    // ...
+        let res = https_agent()
```
```suggestion
    let res = https_agent()
```

nit fmt

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,359 @@
+// pub fn try_payjoin(pj_uri: payjoin::Uri<'static, NetworkChecked>) -> Txid {
+    // ...
+        let res = https_agent()
+        .post(req.url.as_str()) // pj_uri.endpoint
+        .header("Content-Type", "text/plain")
+        .body(String::from_utf8(req.body).unwrap())
```
```suggestion
        .post(req.url) // pj_uri.endpoint
        .header("Content-Type", "text/plain")
        .body(req.body)
```
The dev kit intends to reduce the need for any HTTP I/O adapter. It should give you appropriate types to pass directly to an HTTP client. If not that might be a bug we can fix.

jbesraa replied:
I accepted this change without testing it. and then I saw your comment. why are you suggesting a code change, with the code you want to see without testing it first?

I spent sometime testing each step and made sure its working. I do expect, if you suggest code changes, to at least validate them locally first?

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,359 @@
+
+##### 3.8 Sign and finalize the Payjoin Proposal PSBT
+
+In the previous step we validated the PSBT and now its time finish the
+transaction, sign the PSBT and broadcast it.
```
```suggestion
In the previous step we validated the PSBT and now it's time finish the
transaction, sign the PSBT and broadcast it.
```

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,359 @@
+// ...
+// pub fn try_payjoin(pj_uri: payjoin::Uri<'static, NetworkChecked>) -> Txid {
+    // ...
+        let psbt = sender_wallet
+        .wallet_process_psbt(&base64::encode(psbt.serialize()), None, None, None)
+        .unwrap()
+        .psbt;
+    let tx = sender_wallet
+        .finalize_psbt(&psbt, Some(true))
+        .unwrap()
+        .hex
+        .unwrap();
+    let txid = sender_wallet.send_raw_transaction(&tx).unwrap();
+    txid
```
```suggestion
    let psbt = sender_wallet
        .wallet_process_psbt(&base64::encode(psbt.serialize()), None, None, None)
        .unwrap()
        .psbt;
    let tx = sender_wallet
        .finalize_psbt(&psbt, Some(true))
        .unwrap()
        .hex
        .unwrap();
    let txid = sender_wallet.send_raw_transaction(&tx).unwrap();
    txid
```

nit fmt

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,359 @@
+
+##### 3.9 And lets call the sender functions in the `main`.
+
+In the code below, `args` also read from the command line to consume the URI
+and utilise `payjoin::Uri::try_from` to convert the string to a valid URI.
```
```suggestion
In the code below, `args` also reads from the command line to consume the URI
and utilizes `payjoin::Uri::try_from` to convert the string to a valid URI.
```

conjugation, spelling

I also notice a switch between URI and Uri. Is that arbitrary, or specific to the spec name vs rust type? If it's specific one could use "URI" and "`Uri`" to make the difference consistent.

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
````diff
@@ -0,0 +1,359 @@
+bitcoin:bcrt1qstujjzput7eznwv27e7qh7lpkgz33fh39g2gpt?amount=12.3&pj=https://localhost:3000/&pjos=0
+```
+
+Copy the payjoin uri as we will need it in the next(final) step.
````
```suggestion
Copy the payjoin uri as we will need it in the next (final) step.
```

I'm familiar with a space outside parens in a written clause being convention. Obviously not in maths/science function calls.
https://www.scribens.com/grammar-rules/typography/spacing.html

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
````diff
@@ -0,0 +1,359 @@
+In the code below, `args` also read from the command line to consume the URI
+and utilise `payjoin::Uri::try_from` to convert the string to a valid URI.
+
+```rust
+fn main() {
+    let args: Vec<String> = std::env::args().collect();
+    let payjoin_uri: String = args[1].parse().unwrap();
+    let payjoin_uri = payjoin::Uri::try_from(payjoin_uri)
+        .unwrap()
+        .assume_checked();
+    let txid = sender::try_payjoin(payjoin_uri);
+    dbg!(&txid);
+}
+```
````
I accidentally included this in the sender module when following along. Perhaps say "Back at the top of the file, replace the existing `main()` function or similar. Easy to fix and somewhat obvious by your lack of inline comments in the rust block, but it was my experience following along.

jbesraa replied:
I actually removed the whole file before I started doing the tutorial, so no "back to top" really. I think everyone does this a bit differently. anyhow, I added a comment to emphasis that we are back to the main function

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
````diff
@@ -0,0 +1,359 @@
+cargo run -- "payjoin-uri"
+```
+
+```shell
+### cargo run -- "bitcoin:bcrt1qstujjzput7eznwv27e7qh7lpkgz33fh39g2gpt?amount=12.3&pj=https://localhost:3000/&pjos=0"
+```
````
I think your instructions here are clear and I appreciate the two code blocks to make the example.

It would also be really cool if either the payjoin tx were pretty printed with an explanation of the inputs/outputs coming from sender and receiver, or if there were instructions on how to view it from regtest cli.

jbesraa replied:
I think we can leave that to a whole separate tutorial. explaining what we do in the inputs and how that is different from normal tx.

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,359 @@
+## Conclusion
+
+We just demonstrated how to integrate `rust-payjoin` with a bitcoin
+wallet, using only the `sender` functionality. You could enable other
```
```suggestion
wallet, using only the [`send` feature module](https://docs.rs/payjoin/latest/payjoin/send/index.html). You could enable other
```

This is a good chance to make it easy to follow a link for those ready to learn more

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,359 @@
+
+We just demonstrated how to integrate `rust-payjoin` with a bitcoin
+wallet, using only the `sender` functionality. You could enable other
+features like `receiver` and `v2` to build more functionality.  We
```
```suggestion
features like [`receive`](https://docs.rs/payjoin/latest/payjoin/receive/index.html) and `v2` to build more functionality.  We
```

there's no `receiver` feature ;)

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,359 @@
+wallet, using only the `sender` functionality. You could enable other
+features like `receiver` and `v2` to build more functionality.  We
+also used the `payjoin-cli` to perform as a receiver, while we could
+also utilise it to be a sender.
```
```suggestion
also utilize it to be a sender.
```

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,359 @@
+also used the `payjoin-cli` to perform as a receiver, while we could
+also utilise it to be a sender.
+
+Thanks for following along.
```
Ideally this article acts as a funnel to get people to use Payjoin Dev Kit in their projects. Consider starting this section with "Thanks for following along." and concluding with "... when you're ready to do more with payjoin check out x and y modules in PDK or using payjoin-cli as a sender" with links.

DanGould reviewed (commented):
Tremendous progress here. The tutorial is a joy to follow and provides explanation for each step along the way. Can't wait to publish 🖊️. Nice writing.

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,359 @@
+// pub fn try_payjoin(pj_uri: payjoin::Uri<'static, NetworkChecked>) -> Txid {
+    // ...
+        let psbt = sender_wallet
+        .wallet_process_psbt(&base64::encode(psbt.serialize()), None, None, None)
```
```suggestion
        .wallet_process_psbt(&psbt.to_string(), None, None, None)
```

`Psbt::to_string()` serializes to base64 encoding

jbesraa commented:
Thanks for the review. I believe I added all of your points. 

I added your handle to the main commit as Co-author as well.
![Screenshot from 2024-03-06 18-21-16](https://github.com/payjoin/payjoindevkit.org/assets/21229375/b6ffb785-62ea-409c-addb-b8b1e59601e0)

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,374 @@
+[Why to use HTTPS in production](https://www.cloudflare.com/learning/ssl/why-is-http-not-secure/)
+
+For Payjoin, it is important to use HTTPS to ensure the privacy of the
+transaction and to prevent potential lose of funds.  Without
```
```suggestion
transaction and to prevent potential loss of funds.  Without
```

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,374 @@
+
+## 4. Receiver
+
+To setup the receiver, we will use `payjoin-cli` a cli payjoin
+implementation developed by the `rust-payjoin` team. We will setup
+this receiver, generate BiP21 URI and go back to the `sender` we
+implemented and give it the URI.
+
+Open your terminal, and:
```
```suggestion
To set up the receiver, we will use `payjoin-cli` a cli payjoin
implementation developed by the `rust-payjoin` team. We will setup
this receiver, generate BIP21 URI and go back to the `sender` we
implemented and give it the URI.

Open your terminal, and:
```

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
````diff
@@ -0,0 +1,374 @@
+
+## 2. Project Setup
+
+Lets set up a new rust project and name it `payjoin-tutorial`
+```shell
````
```suggestion
Lets set up a new rust project and name it `payjoin-tutorial`

```shell
```

nit

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,374 @@
+
+##### 3.4 Construct an "Original PSBT" paying `.amount` to `.address`
+
+as specified in the URI.
```
I know I suggested that a space be added after headings, but perhaps this line is this actually meant to be part of the heading?

It seems you're using automatic newlines, and I don't believe markdown will render headings on multiple lines. When I write markdown I just have my editor wordwrap and write long lines.

jbesraa replied:
yea it was a mistaken line-fixing. 

removed totally the .amount and .address from header as they are mentioned in the body just next to it.

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,374 @@
+
+## 4. Receiver
+
+To setup the receiver, we will use `payjoin-cli` a cli payjoin
+implementation developed by the `rust-payjoin` team. We will setup
+this receiver, generate BiP21 URI and go back to the `sender` we
+implemented and give it the URI.
```
```suggestion
To setu p the receiver, we will use `payjoin-cli` a cli payjoin
implementation developed by the `rust-payjoin` team. We will setup
this receiver, generate BIP21 URI and go back to the `sender` we
implemented and give it the URI.
```

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,373 @@
+
+In order to perform a payjoin transaction, you need to have two wallets. 
+lets use [regtest-util](https://github.com/jbesraa/regtest-util) to
+setup the bitcoin wallets.
```
```suggestion
set up the bitcoin wallets.
```
```suggestion
setup the bitcoin wallets.
```

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,373 @@
+
+In order to send a payjoin transaction, the sender needs to complete two steps.
+
+First, the sender obtains the receiver [BIP21 Bitcoin URI](https://github.com/bitcoin/bips/blob/master/bip-0021.mediawiki) (often via QR code) and construct a PSBT
```
```suggestion
First, the sender obtains the receiver [BIP21 Bitcoin URI](https://github.com/bitcoin/bips/blob/master/bip-0021.mediawiki) (often via QR code) and constructs a PSBT
```

DanGould commented on docs/tutorials/send-payjoin-with-pdk.md:
```diff
@@ -0,0 +1,373 @@
+##### 3.9 And lets call the sender functions in the `main`.
+
+Back to the `main` function, we will read the payjoin uri from the
+command line arguments and utilize `payjoin::Uri` to parse it.
```
```suggestion
command line arguments and utilize `payjoin::Uri` to parse them.
```

parse [the arguments]

DanGould reviewed (approved):
This is good to go. I found a few more nits but I still approve.
